content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+
+## Design Goals
+The authors provide two design goals for Csmith:
+1. Every randomly generated program must be well formed and have a single interpretation based on the C standard.
+2. Maximize "expressiveness". Expressiveness is the idea that the generated programs should use a wide variety and combinations of language features
+   - For example, Csmith allows for programs with function definitions, global definitions, most C expressions and statements, control flow, arrays, pointers, and more. Some language features that are disallowed include dynamic memory allocation, recursion, and function pointers.
+
+Interestingly, design goal #2 is in direct opposition to design goal #1. The more features we allow to be tested (like arrays and pointers), the easier it is to cause undefined behavior. First, we will discuss at a high level how programs are randomly generated. Then, we will explore how Csmith ensures the generated code doesn't exhibit undefined behavior.
+
+## Randomly Generating Programs
+To generate programs, first Csmith generates types that can be used later during program generation. Then starting from `main`, the authors break down how the program gets filled in.
+1. Csmith selects a random production from the grammar, based on the current context, with potentially non-uniform probability. For example, when inside a function body, Csmith can choose things like `if` and `for` statements, but not a `continue` if the code is not in a loop.
+    - Csmith operates on a subset of C. Interestingly, it doesn't seem like the authors have a grammar explicitly written down for this subset.
+    - For productions that require a *target* (e.g. a variable), Csmith can randomly decide to generate a new target or use an existing one.
+    - Similarly goes for productions requiring a type.
+2. When choosing a nonterminal production, Csmith will recursively choose productions given the appropriate context. For example, if a function call production is chosen, then expressions for the parameters must be generated.
+3. Csmith implements an interprocedural pointer analysis, but needs to keep track of "points-to facts" about the program. As the program is being generated, the set of facts are being updated.
+4. Csmith uses a set of *saftey checks* to make sure the newly generated bit of code is well formed, following design goal #1. If a safety check fails, the changes are undone and we try again.
+
+TODO: Provide an example.
+
+## Safety Mechanisms
+
+A program can crash for various reasons. It might be because of compiler failures, but more often, due to unsafe programs. Csmith wants to avoid these pitfalls with proper safety mechanisms.
+
+**Discussion:** Think about what safety issues can be caused by the randomly generated programs.
```

Comment:
Your earlier discussion question was in italics, and it was labeled "Discussion Question:" instead of "Discussion:". Consider picking a style and staying consistent.